File: app/middleware/logging.py
"""
请求日志中间件

记录所有HTTP请求的详细信息
"""

import time
import uuid
from typing import Callable

from fastapi import Request, Response
from starlette.middleware.base import BaseHTTPMiddleware

from app.core.logging import bind_request_id, get_logger

logger = get_logger(__name__)


class LoggingMiddleware(BaseHTTPMiddleware):
    """
    请求日志中间件

    记录每个HTTP请求的详细信息，包括：
    - 请求ID
    - 请求方法、路径
    - 请求处理时间
    - 响应状态码
    """

    async def dispatch(self, request: Request, call_next: Callable) -> Response:
        """
        处理请求并记录日志

        Args:
            request: HTTP请求对象
            call_next: 下一个中间件或路由处理函数

        Returns:
            Response: HTTP响应对象
        """
        # 生成请求ID
        request_id = str(uuid.uuid4())
        bind_request_id(request_id)

        # 记录请求开始时间
        start_time = time.time()

        # 记录请求信息
        logger.info(
            "请求开始",
            method=request.method,
            path=request.url.path,
            query_params=str(request.query_params),
            client_host=request.client.host if request.client else None,
            user_agent=request.headers.get("user-agent"),
        )

        try:
            # 处理请求
            response = await call_next(request)

            # 计算处理时间
            process_time = time.time() - start_time

            # 记录响应信息
            logger.info(
                "请求完成",
                method=request.method,
                path=request.url.path,
                status_code=response.status_code,
                process_time=f"{process_time:.3f}s",
            )

            # 添加请求ID到响应头
            response.headers["X-Request-ID"] = request_id
            response.headers["X-Process-Time"] = f"{process_time:.3f}"

            return response

        except Exception as e:
            # 计算处理时间
            process_time = time.time() - start_time

            # 记录错误信息
            logger.error(
                "请求失败",
                method=request.method,
                path=request.url.path,
                error=str(e),
                error_type=type(e).__name__,
                process_time=f"{process_time:.3f}s",
                exc_info=True,
            )

            raise

File: src/middleware/logging.ts
import { FastifyPluginAsync, FastifyRequest, FastifyReply } from "fastify";
import { randomUUID } from "crypto";
import { getLogger } from "../core/logger.js";
import { settings } from "../core/config.js";
import { t, type Locale } from "../i18n/index.js";

const logger = getLogger("request");
const logLocale = (): Locale => settings.DEFAULT_LOCALE as Locale;

/**
 * 扩展 Fastify 请求类型，添加请求 ID 与开始时间（用于计算响应耗时）
 */
declare module "fastify" {
  interface FastifyRequest {
    requestId: string;
    requestStartTime?: number;
  }
}

/**
 * 请求日志中间件插件
 *
 * 记录每个 HTTP 请求的详细信息，包括：
 * - 请求ID
 * - 请求方法、路径
 * - 请求处理时间
 * - 响应状态码
 */
export const loggingPlugin: FastifyPluginAsync = async (fastify) => {
  // 添加请求 ID 到请求对象
  fastify.addHook("onRequest", async (request: FastifyRequest) => {
    request.requestId = randomUUID();
    request.requestStartTime = Date.now();

    logger.info({
      message: t("log.request.start", undefined, logLocale()),
      requestId: request.requestId,
      method: request.method,
      path: request.url,
      query: request.query,
      ip: request.ip,
      userAgent: request.headers["user-agent"],
    });
  });

  // 记录响应信息
  fastify.addHook("onResponse", async (
    request: FastifyRequest,
    reply: FastifyReply
  ) => {
    const elapsed =
      request.requestStartTime != null ? Date.now() - request.requestStartTime : 0;
    const responseTimeMs = elapsed;

    logger.info({
      message: t("log.request.complete", undefined, logLocale()),
      requestId: request.requestId,
      method: request.method,
      path: request.url,
      statusCode: reply.statusCode,
      responseTime: `${responseTimeMs.toFixed(3)}ms`,
    });

    reply.header("X-Request-ID", request.requestId);
    reply.header("X-Process-Time", responseTimeMs.toFixed(3));
  });

  // 记录错误信息
  fastify.addHook("onError", async (
    request: FastifyRequest,
    reply: FastifyReply,
    error: Error
  ) => {
    const elapsed =
      request.requestStartTime != null ? Date.now() - request.requestStartTime : 0;
    const responseTimeMs = elapsed;

    logger.error({
      message: t("log.request.failed", undefined, logLocale()),
      requestId: request.requestId,
      method: request.method,
      path: request.url,
      statusCode: reply.statusCode,
      responseTime: `${responseTimeMs.toFixed(3)}ms`,
      error: error.message,
      errorType: error.constructor.name,
      stack: error.stack,
    });
  });
};
